cmp/compare.go

```diff
@@ -361,26 +352,55 @@ func (s *state) tryMethod(vx, vy reflect.Value, t reflect.Type) bool {
 		return false
 	}
 
-	eq := s.callFunc(m.Func, vx, vy)
+	eq := s.callTTBFunc(m.Func, vx, vy)
 	s.report(eq, vx, vy)
 	return true
 }
 
-func (s *state) callFunc(f, x, y reflect.Value) bool {
-	got := f.Call([]reflect.Value{x, y})[0].Bool()
-	if s.dsCheck.curr == s.dsCheck.next {
-		// Swapping the input arguments is sufficient to check that
-		// f is symmetric and deterministic.
-		want := f.Call([]reflect.Value{y, x})[0].Bool()
-		if got != want {
-			fn := getFuncName(f.Pointer())
-			panic(fmt.Sprintf("non-deterministic or non-symmetric function detected: %s", fn))
-		}
-		s.dsCheck.curr = 0
-		s.dsCheck.next++
+func (s *state) callTRFunc(f, v reflect.Value) reflect.Value {
+	if !s.dynChecker.Next() {
+		return f.Call([]reflect.Value{v})[0]
+	}
+
+	// Run the function twice and ensure that we get the same results back.
+	// We run in goroutines so that the race detector (if enabled) can detect
+	// unsafe mutations to the input.
+	c := make(chan reflect.Value)
+	go detectRaces(c, f, v)
+	want := f.Call([]reflect.Value{v})[0]
+	if got := <-c; !value.Equal(got, want) {
```

Comment:
In an implementation of lazy-evaluated value, wouldn't the overall structure of the value returned still be the same (i.e., the returned value in both cases are in the lazy form)? In the case of a memoized value, would it be the exact same value returned from each call to the transformer? Either way, our loose definition of DeepEqual would work. For the detection of mutation on the *input*, extensions on proto messages was a realistic case where it would fail. What's a realistic case where this check on the *output* would fail? 

If possible I want to avoid plumbing this through all the logic of `compareAny`, which would require needing to ensure the `curPath` is updated, that we don't accidentally output to the reporter, and also that we don't clobber any other state.